lib/commons/dom/is-visible.js

```diff
@@ -179,56 +89,194 @@ function isVisible(el, screenReader, recursed) {
     return isAreaVisible(el, screenReader, recursed);
   }
 
-  // always hidden
-  if (style.getPropertyValue('display') === 'none') {
+  // different ways to hide an element
+  const hiddenMethods = [
+    displayHidden,
+    opacityHidden,
+    clipHidden,
+    scrollHidden,
+    overflowHidden,
+    visiblityHidden,
+    offscreenHidden
+  ];
```

Comment:
I think you should preselect these methods based on the current scenario, rather than pass `recursed` and `screenreader` into these methods:

```suggestion
  const hiddenMethods = sharedHiddenMethods.concat(
    screenReader ? accTreeHiddenMethods : offScreenMethods
  )
  if (!screenReader && !recursed) {
    hiddenMethods.push(...nonRecursedScreenMethods);
  }
```

Reply:
So I tried this and I'm not sure I like the approach. Because of the divergent code paths with disconnected trees (which can check ariaHidden and elementHidden, but nothing else), the final code ended up looking something like this:

```js
let hiddenMethods = [elementHidden];
let style;
if (screenReader) {
  hiddenMethods.push(ariaHidden)
}

if (el) {
  if (nodeType === DOCUMENT_FRAGMENT_NODE) {
    el = el.host; // grab the host Node
  }

  style = window.getComputedStyle(el, null);
  if (style === null) {
    return false;
  }

  /**
   * check visibility of `AREA`
   * Note:
   * Firefox's user-agent always sets `AREA` element to `display:none`
   * hence excluding the edge case, for visibility computation
   */
  if (nodeName === 'area') {
    return isAreaVisible(el, screenReader, recursed);
  }

  hiddenMethods.push(displayHidden);

  if (!screenReader) {
    hiddenMethods.push(
      opacityHidden,
      clipHidden,
      scrollHidden,
      overflowHidden
    );
  }
  if (!recursed) {
    hiddenMethods.push(visibilityHidden);
  }
  if (!screenReader && !recursed) {
    hiddenMethods.push(offscreenHidden);
  }
}

if (
  hiddenMethods.some(method =>
    method({
      el,
      vNode,
      nodeName,
      style
    })
  )
) {
  return false;
}
```

I find it harder to see what's going on, and even harder to know which methods are allowed to look at `vNode` and which must stick to just using the `el`. Also that style is passed to all methods but isn't set unless you have `el` is kinda odd to me.